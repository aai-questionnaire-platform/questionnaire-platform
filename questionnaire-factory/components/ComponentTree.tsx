import dynamic from 'next/dynamic';
import { createElement, memo } from 'react';

import type { ComponentDefinition } from '@/schema/Components';
import { uniqueId } from '@/util';

const COMPONENTS: Record<string, any> = {
  __TEST_ONLY_TYPOGRAPHY: dynamic(() => import('../components/Typography')),
  FRONT_PAGE: dynamic(() => import('../features/FrontPage')),
  CATEGORY: dynamic(() => import('../features/Category')),
  CATEGORY_END: dynamic(() => import('../features/CategoryEnd')),
  PROGRESS: dynamic(() => import('../features/Progress')),
  ANSWER_SUMMARY: dynamic(() => import('../features/AnswerSummary')),
  FEEDBACK: dynamic(() => import('../features/Feedback')),
  ADMIN_GROUPS: dynamic(() => import('../features/AdminGroups')),
  REGISTRATION: dynamic(() => import('../features/Registration')),
  ADMIN_CATEGORIES: dynamic(() => import('../features/AdminCategories')),
  ADMIN_ANSWER_SUMMARY: dynamic(() => import('../features/AdminAnswerSummary')),
  HN_CATEGORY: dynamic(() => import('../features/HnCategory')),
  HN_RESULTS: dynamic(() => import('../features/HnResults')),
  HN_FRONT_PAGE: dynamic(() => import('../features/HnFrontPage')),
  HN_SIGN_IN_HANDLER: dynamic(
    () => import('../features/HnResults/HnSignInHandler')
  ),
};

function renderComponent(component: ComponentDefinition) {
  if (typeof component === 'string') {
    return component;
  }

  const Component = COMPONENTS[component.type];

  return createElement(
    Component,
    { ...component.props, key: uniqueId(component.type) },
    // if the component has children then recursively render them too
    Array.isArray(component.children)
      ? component.children.map(renderComponent as any)
      : component.children ?? null
  );
}

interface ComponentTreeProps {
  tree?: ComponentDefinition[];
}

/**
 * Component responsible for rendering the ui from the app structure json
 * @param props
 * @returns
 */
const ComponentTree = memo(({ tree = [] }: ComponentTreeProps) => (
  <>{tree.map(renderComponent)}</>
));

ComponentTree.displayName = 'ComponentTree';

export default ComponentTree;
