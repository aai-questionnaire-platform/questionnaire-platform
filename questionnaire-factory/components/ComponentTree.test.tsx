import { render } from '@testing-library/react';
import preloadAll from 'jest-next-dynamic';
import 'jest-styled-components';

import ComponentTree from '@/components/ComponentTree';

beforeAll(async () => {
  await preloadAll();
});

it('should render an empty tree', () => {
  expectToMatchSnapshotWithTree([]);
});

it('should render a string', () => {
  expectToMatchSnapshotWithTree(['string child']);
});

it('should render a component without children', () => {
  expectToMatchSnapshotWithTree([
    {
      type: '__TEST_ONLY_TYPOGRAPHY',
      props: { variant: 'default' },
    },
  ]);
});

it('should render a component with string as a child', async () => {
  expectToMatchSnapshotWithTree([
    {
      type: '__TEST_ONLY_TYPOGRAPHY',
      props: { variant: 'default' },
      children: ['child string'],
    },
  ]);
});

it('should render a component with component as a child', async () => {
  expectToMatchSnapshotWithTree([
    {
      type: '__TEST_ONLY_TYPOGRAPHY',
      props: { variant: 'default' },
      children: [
        {
          type: '__TEST_ONLY_TYPOGRAPHY',
          props: { as: 'span', variant: 'default' },
          children: 'grandchild text',
        },
      ],
    },
  ]);
});

it('should render a tree with siblings', () => {
  expectToMatchSnapshotWithTree([
    {
      type: '__TEST_ONLY_TYPOGRAPHY',
      props: { variant: 'default' },
      children: 'sibling 1',
    },
    {
      type: '__TEST_ONLY_TYPOGRAPHY',
      props: { variant: 'small' },
      children: 'sibling 2',
    },
  ]);
});

it('should render a tree with siblings with children', () => {
  expectToMatchSnapshotWithTree([
    {
      type: '__TEST_ONLY_TYPOGRAPHY',
      props: { variant: 'default' },
      children: [
        {
          type: '__TEST_ONLY_TYPOGRAPHY',
          props: { as: 'span', variant: 'default' },
          children: 'grandchild 1',
        },
      ],
    },
    {
      type: '__TEST_ONLY_TYPOGRAPHY',
      props: { variant: 'small' },
      children: [
        {
          type: '__TEST_ONLY_TYPOGRAPHY',
          props: { as: 'span', variant: 'small' },
          children: 'grandchild 2',
        },
      ],
    },
  ]);
});

function expectToMatchSnapshotWithTree(tree: any) {
  const { container } = render(<ComponentTree tree={tree} />);
  expect(container).toMatchSnapshot();
}
