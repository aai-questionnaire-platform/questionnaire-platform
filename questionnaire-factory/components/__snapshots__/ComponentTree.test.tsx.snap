// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a component with component as a child 1`] = `
.c0 {
  font-size: 1rem;
  line-height: 1.5rem;
  white-space: pre-line;
}

<div>
  <p
    class="c0"
  >
    <span
      class="c0"
    >
      grandchild text
    </span>
  </p>
</div>
`;

exports[`should render a component with string as a child 1`] = `
.c0 {
  font-size: 1rem;
  line-height: 1.5rem;
  white-space: pre-line;
}

<div>
  <p
    class="c0"
  >
    child string
  </p>
</div>
`;

exports[`should render a component without children 1`] = `
.c0 {
  font-size: 1rem;
  line-height: 1.5rem;
  white-space: pre-line;
}

<div>
  <p
    class="c0"
  />
</div>
`;

exports[`should render a string 1`] = `
<div>
  string child
</div>
`;

exports[`should render a tree with siblings 1`] = `
.c0 {
  font-size: 1rem;
  line-height: 1.5rem;
  white-space: pre-line;
}

.c1 {
  font-size: 0.875rem;
  line-height: 1.25rem;
  white-space: pre-line;
}

<div>
  <p
    class="c0"
  >
    sibling 1
  </p>
  <p
    class="c1"
  >
    sibling 2
  </p>
</div>
`;

exports[`should render a tree with siblings with children 1`] = `
.c0 {
  font-size: 1rem;
  line-height: 1.5rem;
  white-space: pre-line;
}

.c1 {
  font-size: 0.875rem;
  line-height: 1.25rem;
  white-space: pre-line;
}

<div>
  <p
    class="c0"
  >
    <span
      class="c0"
    >
      grandchild 1
    </span>
  </p>
  <p
    class="c1"
  >
    <span
      class="c1"
    >
      grandchild 2
    </span>
  </p>
</div>
`;

exports[`should render an empty tree 1`] = `<div />`;
